Telegram/SourceFiles/export/data/export_data_types.cpp

```diff
@@ -318,6 +318,77 @@ std::vector<TextPart> ParseText(
 	return result;
 }
 
+Utf8String Reaction::TypeToString(const Reaction &reaction) {
+	switch (reaction.type) {
+		case Reaction::Type::Empty: return "empty";
+		case Reaction::Type::Emoji: return "emoji";
+		case Reaction::Type::CustomEmoji: return "custom_emoji";
+	}
+	Unexpected("Type in Reaction::Type.");
+}
+
+Utf8String Reaction::Id(const Reaction &reaction) {
+	Utf8String id;
+	switch (reaction.type) {
+		case Reaction::Type::Emoji:
+			id = reaction.emoji.toUtf8();
+			break;
+		case Reaction::Type::CustomEmoji:
+			id = reaction.documentId;
+			break;
+	}
+	return Reaction::TypeToString(reaction) + id;
+}
+
+Reaction ParseReaction(const MTPReaction& reaction) {
+	Reaction result;
+	reaction.match([&](const MTPDreactionEmoji &data) {
+		result.type = Reaction::Type::Emoji;
+		result.emoji = qs(data.vemoticon());
+	}, [&](const MTPDreactionCustomEmoji &data) {
+		result.type = Reaction::Type::CustomEmoji;
+		result.documentId = NumberToString(data.vdocument_id().v);
+	}, [&](const MTPDreactionPaid &data) {
+		result.type = Reaction::Type::Paid;
+	}, [&](const MTPDreactionEmpty &data) {
+		result.type = Reaction::Type::Empty;
+	});
+	return result;
+}
+
+std::vector<Reaction> ParseReactions(const MTPMessageReactions &data) {
+	std::map<QString, Reaction> reactionsMap;
+	std::vector<Utf8String> reactionsOrder;
+	for (const auto &single : data.data().vresults().v) {
+		Reaction reaction = ParseReaction(single.data().vreaction());
+		reaction.count = single.data().vcount().v;
+		Utf8String id = Reaction::Id(reaction);
+		auto const &[_, inserted] = reactionsMap.try_emplace(id, reaction);
+		if (inserted) {
+			reactionsOrder.push_back(id);
+		}
+	}
+	if (data.data().vrecent_reactions().has_value()) {
+		for (const auto &single : data.data().vrecent_reactions().value_or_empty()) {
+			Reaction reaction = ParseReaction(single.data().vreaction());
+			Utf8String id = Reaction::Id(reaction);
+			auto const &[it, inserted] = reactionsMap.try_emplace(id, reaction);
+			if (inserted) {
+				reactionsOrder.push_back(id);
+			}
+			it->second.recent.push_back({
+				.peerId = ParsePeerId(single.data().vpeer_id()),
+				.date = single.data().vdate().v,
+			});
+		}
+	}
+	std::vector<Reaction> results;
+	for (const auto& id : reactionsOrder) {
```

Comment:
The code style of this repository demands `const auto &id` instead of `const auto& id`.
```suggestion
	for (const auto &id : reactionsOrder) {
```